packages/@o3r/schematics/src/utility/tree.ts

```diff
@@ -0,0 +1,23 @@
+import type { DirEntry, FileEntry } from '@angular-devkit/schematics';
+
+/**
+ *
+ * Helper function that looks for files in the Tree
+ *
+ * @param memory where to store the results
+ * @param directory where to perform the search
+ * @param fileMatchesCriteria a function defining the criteria to look for
+ * @param ignoreDirectories optional parameter to ignore folders
+ */
+export function findFilenameInTree(memory: Set<FileEntry>, directory: DirEntry, fileMatchesCriteria: (file: string) => boolean, ignoreDirectories: string[] = ['node_modules', '.git', '.yarn']) {
```

Comment:
Never export a recursive function with a memory. You should export a function without memory using the recursive function

Reply:
how about 
```typescript
import type { DirEntry, FileEntry } from '@angular-devkit/schematics';

/**
 *
 * Helper function that looks for files in the Tree
 *
 * @param directory where to perform the search
 * @param fileMatchesCriteria a function defining the criteria to look for
 * @param ignoreDirectories optional parameter to ignore folders
 */
export function findFilenameInTree(directory: DirEntry, fileMatchesCriteria: (file: string) => boolean, ignoreDirectories: string[] = ['node_modules', '.git', '.yarn']): Set<FileEntry> {
  const memory = new Set<FileEntry>();

  function recursiveSearch(directory: DirEntry): void {
    if (ignoreDirectories.some(i => directory.path.includes(i))) {
      return;
    }

    directory.subfiles
      .filter(fileMatchesCriteria)
      .forEach((file) => memory.add(directory.file(file)!));

    directory.subdirs
      .forEach((dir) => recursiveSearch(directory.dir(dir)));
  }

  recursiveSearch(directory);

  return memory;
}
`
``

Reply:
In term of perf: you instancy an impure rec function at each call, you can improve it to simply:

```typescript
function findFilenameInTreeRec(memory: Set<FileEntry>, directory: DirEntry, fileMatchesCriteria: (file: string) => boolean, ignoreDirectories: string[] = ['node_modules', '.git', '.yarn']) {
  if (ignoreDirectories.some(i => directory.path.includes(i))) {return memory;}

  directory.subfiles
    .filter(fileMatchesCriteria)
    .forEach((file) => memory.add(directory.file(file)!));

  directory.subdirs
    .forEach((dir) => findFilenameInTree(memory, directory.dir(dir), fileMatchesCriteria, ignoreDirectories));

  return memory;
}

export function fndFilenameInTree(directory: DirEntry, fileMatchesCriteria: (file: string) => boolean, ignoreDirectories: string[] = ['node_modules', '.git', '.yarn']) {
  const memory = new Set<FileEntry>();
  return [...findFilenameInTreeRec(directory, fileMatchesCriteria)];
```